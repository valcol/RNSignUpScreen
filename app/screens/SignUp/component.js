import React, { useState, useRef, useEffect } from "react";
import {
    Body,
    Button,
    Card,
    Container,
    Form,
    Icon,
    Left,
    Title,
    Toast
} from "native-base";
import {
    ButtonsWrapper,
    CustomButton,
    DigitInput,
    DigitInputItem,
    ErrorCardContent,
    ErrorCardIcon,
    ErrorCardItem,
    FormItem,
    KeyboardAwareWrapper,
    PhoneInput,
    PhoneInputItem,
    ScreenContent,
    SMSCodeContainer,
    ThemedHeader,
    WhiteLabel,
    SubmitPhoneNumberIcon
} from "./styles";
import R from "ramda";
import withSignUp from "../../store/signup/hoc";

const SignUp = ({
    isCodeRequestPending,
    isCodeRequestOnError,
    isCodeRequestOnSuccess,
    isCodeVerificationPending,
    isCodeVerificationOnError,
    isCodeVerificationOnSuccess,
    requestSMSCode,
    verifySMSCode
}) => {
    const [phoneNumber, setPhoneNumber] = useState("");
    const [SMSCode, setSMSCode] = useState(["", "", "", ""]);
    const [isPhoneNumberInvalid, setIsPhoneNumberInvalid] = useState(
        isCodeRequestOnError
    );
    const [isSMSCodeInvalid, setIsSMSCodeInvalid] = useState(
        isCodeVerificationOnError
    );
    const digitInputRefs = [useRef(), useRef(), useRef(), useRef()];
    useEffect(() => setIsPhoneNumberInvalid(isCodeRequestOnError), [
        isCodeRequestOnError
    ]);
    useEffect(() => setIsSMSCodeInvalid(isCodeVerificationOnError), [
        isCodeVerificationOnError
    ]);
    useEffect(() => {
        if (!isCodeVerificationOnSuccess) return;
        Toast.show({
            text: "Success!",
            buttonText: "Okay",
            type: "success"
        });
    }, [isCodeVerificationOnSuccess]);

    const setSMSCodeDigit = (index, digit) =>
        setSMSCode(R.set(R.lensIndex(index), digit, SMSCode));

    const isPhoneNumberFilled = phoneNumber.length == 10;

    const isSMSCodeFilled = !SMSCode.includes("");

    const focusNextInput = currentInputIndex => {
        const nextInputIndex = currentInputIndex + 1;
        if (nextInputIndex > SMSCode.length) return;

        const nextInputRef = R.pathOr(null, [nextInputIndex], digitInputRefs);
        if (!nextInputRef) return;

        nextInputRef.current._root.focus();
    };

    const submitCode = () => verifySMSCode(SMSCode.join(""));

    const submitPhoneNumber = () => requestSMSCode(phoneNumber);

    const renderSMSCodeDigitField = index => (
        <DigitInputItem regular error={isSMSCodeInvalid}>
            <DigitInput
                blurOnSubmit={false}
                keyboardType={"phone-pad"}
                maxLength={1}
                onChangeText={digit => {
                    setSMSCodeDigit(index, digit);
                    focusNextInput(index);
                    setIsSMSCodeInvalid(false);
                }}
                onFocus={() => setSMSCodeDigit(index, "")}
                value={SMSCode[index]}
                ref={digitInputRefs[index]}
            />
        </DigitInputItem>
    );

    const renderSubmitPhoneNumberButton = () => {
        if (!isPhoneNumberFilled) return null;

        return (
            <Button transparent onPress={submitPhoneNumber}>
                <SubmitPhoneNumberIcon name="send" />
            </Button>
        );
    };

    const renderSMSCodeForm = () => {
        if (!isCodeRequestOnSuccess) return null;

        return (
            <FormItem>
                <WhiteLabel>Code de confirmation reçu par SMS</WhiteLabel>
                <SMSCodeContainer>
                    {renderSMSCodeDigitField(0)}
                    {renderSMSCodeDigitField(1)}
                    {renderSMSCodeDigitField(2)}
                    {renderSMSCodeDigitField(3)}
                </SMSCodeContainer>
                {isSMSCodeInvalid && renderError("Code incorrect.")}
            </FormItem>
        );
    };

    const renderError = message => (
        <Card>
            <ErrorCardItem>
                <ErrorCardContent>
                    <ErrorCardIcon name="close-circle" />
                    <WhiteLabel>{message}</WhiteLabel>
                </ErrorCardContent>
            </ErrorCardItem>
        </Card>
    );

    const renderResendButton = () => {
        if (!isCodeRequestOnSuccess) return null;

        return (
            <CustomButton rounded bordered onPress={submitPhoneNumber}>
                <WhiteLabel>Renvoyer le SMS</WhiteLabel>
            </CustomButton>
        );
    };

    return (
        <Container>
            <ThemedHeader>
                <Left>
                    <Button transparent>
                        <Icon name="arrow-back" />
                    </Button>
                </Left>
                <Body>
                    <Title>Créez votre compte</Title>
                </Body>
            </ThemedHeader>
            <KeyboardAwareWrapper behavior="height">
                <ScreenContent>
                    <Form>
                        <FormItem>
                            <WhiteLabel>Numéro de téléphone</WhiteLabel>
                            <PhoneInputItem
                                regular
                                error={isPhoneNumberInvalid}
                            >
                                <PhoneInput
                                    onChangeText={number => {
                                        setPhoneNumber(number);
                                        setIsPhoneNumberInvalid(false);
                                    }}
                                    value={phoneNumber}
                                    keyboardType={"phone-pad"}
                                    maxLength={10}
                                />
                                {renderSubmitPhoneNumberButton()}
                            </PhoneInputItem>
                            {isPhoneNumberInvalid &&
                                renderError("Numéro incorrect.")}
                        </FormItem>
                        {renderSMSCodeForm()}
                    </Form>
                    <ButtonsWrapper>
                        {renderResendButton()}
                        <CustomButton
                            rounded
                            disabled={!isSMSCodeFilled}
                            onPress={submitCode}
                        >
                            <WhiteLabel>Suivant</WhiteLabel>
                        </CustomButton>
                    </ButtonsWrapper>
                </ScreenContent>
            </KeyboardAwareWrapper>
        </Container>
    );
};

export default withSignUp(SignUp);
