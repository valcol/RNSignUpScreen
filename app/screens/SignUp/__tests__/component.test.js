import "react-native";
import React from "react";
import { SignUp } from "../component";
import renderer from "react-test-renderer";

it("renders correctly with no props", () => {
    const tree = renderer.create(<SignUp />).toJSON();
    expect(tree).toMatchSnapshot();
});
