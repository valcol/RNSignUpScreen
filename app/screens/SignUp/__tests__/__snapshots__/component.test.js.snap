// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with no props 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "height": 1334,
    }
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#F8F8F8",
            "borderBottomColor": "#a7a6ab",
            "borderBottomWidth": 0.5,
            "elevation": 3,
            "flexDirection": "row",
            "height": 64,
            "justifyContent": "center",
            "left": 0,
            "paddingLeft": 6,
            "paddingRight": 10,
            "paddingTop": 18,
            "right": 0,
            "shadowColor": undefined,
            "shadowOffset": undefined,
            "shadowOpacity": undefined,
            "shadowRadius": undefined,
            "top": 0,
          },
          Object {
            "backgroundColor": "#01a2d9",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "center",
            "flex": 1,
          }
        }
      >
        <View
          accessible={true}
          clickable={false}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": null,
              "backgroundColor": "transparent",
              "borderBottomWidth": null,
              "borderColor": null,
              "borderLeftWidth": null,
              "borderRadius": 5,
              "borderRightWidth": null,
              "borderTopWidth": null,
              "borderWidth": undefined,
              "elevation": 0,
              "flexDirection": "row",
              "height": 45,
              "justifyContent": "space-between",
              "marginLeft": -3,
              "opacity": 1,
              "paddingBottom": 6,
              "paddingLeft": 4,
              "paddingRight": 6,
              "paddingTop": 6,
              "shadowColor": null,
              "shadowOffset": null,
              "shadowOpacity": null,
              "shadowRadius": null,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#007aff",
                    "fontSize": 34,
                    "marginLeft": 1,
                    "marginRight": 2,
                    "marginTop": 0,
                    "paddingTop": 1,
                  },
                ],
                Object {
                  "fontFamily": "Ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "flex": 1,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 17,
              "fontWeight": "700",
              "marginLeft": undefined,
              "paddingLeft": 4,
              "paddingTop": 1,
              "textAlign": "center",
            }
          }
        >
          Créez votre compte
        </Text>
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#0084b2",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "space-between",
            "paddingBottom": 20,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 20,
          },
        ]
      }
    >
      <View
        style={Object {}}
      >
        <View
          style={
            Array [
              Object {
                "height": 140,
                "marginBottom": 10,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 10,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 17,
                },
                Object {
                  "color": "white",
                },
              ]
            }
          >
            Numéro de téléphone
          </Text>
          <View
            accessible={true}
            clickable={false}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 1,
                "borderColor": "#D9D5DC",
                "borderLeftWidth": 1,
                "borderRadius": 5,
                "borderRightWidth": 1,
                "borderTopWidth": 1,
                "borderWidth": undefined,
                "flexDirection": "row",
                "marginLeft": 0,
                "opacity": 1,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              editable={true}
              keyboardType="phone-pad"
              maxLength={10}
              onChangeText={[Function]}
              placeholderTextColor="#575757"
              rejectResponderTermination={true}
              style={
                Array [
                  Object {
                    "color": "#000",
                    "flex": 1,
                    "fontSize": 17,
                    "height": 50,
                    "paddingLeft": 8,
                    "paddingRight": 5,
                    "top": 1.5,
                  },
                  Object {
                    "color": "white",
                  },
                ]
              }
              underlineColorAndroid="rgba(0,0,0,0)"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "flex-end",
            },
          ]
        }
      >
        <View
          accessible={true}
          clickable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "grey",
              "borderBottomWidth": null,
              "borderColor": "#01a2d9",
              "borderLeftWidth": null,
              "borderRadius": 57,
              "borderRightWidth": null,
              "borderTopWidth": null,
              "borderWidth": undefined,
              "elevation": 2,
              "flexDirection": "row",
              "height": 45,
              "justifyContent": "space-between",
              "marginBottom": 5,
              "marginLeft": 5,
              "marginRight": 5,
              "marginTop": 5,
              "opacity": 1,
              "paddingBottom": 0,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 0,
              "shadowColor": undefined,
              "shadowOffset": undefined,
              "shadowOpacity": undefined,
              "shadowRadius": undefined,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 17,
                },
                Object {
                  "color": "white",
                },
              ]
            }
          >
            Suivant
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
